vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/pool/SqlConnectionPool.java

```diff
@@ -0,0 +1,245 @@
+  private final int maxSize;
+  private boolean closed;
+
+  public SqlConnectionPool(ConnectionFactory connector, int maxSize) {
+    this(connector, maxSize, PoolOptions.DEFAULT_MAX_WAIT_QUEUE_SIZE);
+  }
+
+  public SqlConnectionPool(ConnectionFactory connector, int maxSize, int maxWaitQueueSize) {
+    this(connector, null, maxSize, maxWaitQueueSize);
+  }
+
+  public SqlConnectionPool(ConnectionFactory factory, Context context, int maxSize, int maxWaitQueueSize) {
+    Objects.requireNonNull(factory, "No null connector");
+    if (maxSize < 1) {
+      throw new IllegalArgumentException("Pool max size must be > 0");
+    }
+    this.pool = new SimpleConnectionPool<>(connector, maxSize, maxSize, maxWaitQueueSize);
+    this.context = (ContextInternal) context;
+    this.maxSize = maxSize;
+    this.factory = factory;
+  }
+
+  private final Connector<PooledConnection> connector = new Connector<PooledConnection>() {
+    @Override
+    public void connect(EventLoopContext context, ConnectionEventListener listener, Handler<AsyncResult<ConnectResult<PooledConnection>>> handler) {
+      PromiseInternal<Connection> promise = context.promise();
+      factory.connect(promise);
+      promise.future()
+        .map(connection -> {
+          PooledConnection pooled = new PooledConnection(connection, listener);
+          connection.init(pooled);
+          return new ConnectResult<>(pooled, 1, 1);
+        })
+        .onComplete(handler);
+    }
+
+    @Override
+    public boolean isValid(PooledConnection connection) {
+      return true;
+    }
+  };
+
+  public int available() {
+    return maxSize - pool.size();
+  }
+
+  public int size() {
+    return pool.size();
+  }
+
+  public void acquire(ContextInternal context, Handler<AsyncResult<Connection>> waiter) {
+    pool.acquire((EventLoopContext) context, 1, ar -> {
+      if (ar.succeeded()) {
+        Lease<PooledConnection> lease = ar.result();
+        PooledConnection pooled = lease.get();
+        pooled.lease = lease;
+        waiter.handle(Future.succeededFuture(pooled));
+      } else {
+        waiter.handle(Future.failedFuture(ar.cause()));
+      }
+    });
+  }
+
+  public Future<Void> close() {
+    PromiseInternal<Void> promise = context.promise();
+    context.emit(promise, this::close);
+    return promise.future();
+  }
+
+  public void close(Promise<Void> promise) {
+    if (closed) {
+      promise.fail("Connection pool already closed");
+      return;
+    }
+    closed = true;
+    List<Future> futures = pool
+      .close()
+      .stream().map(f -> f
+        .flatMap(c -> {
+          Promise<Void> p = Promise.promise();
+          c.close(p);
+          return p.future();
+        }))
+      .collect(Collectors.toList());
+    CompositeFuture
+      .join(futures)
+      .<Void>mapEmpty()
+      .onComplete(promise);
+  }
+
+  private class PooledConnection implements Connection, Connection.Holder  {
+
+    private final Connection conn;
```

Comment:
I'm wondering whether we can remove `Connection` interface and just wrap `SqlConnection` in `PooledConnection` instead as we have to wrap `Connection` in `SqlConnection` each time for connection acquirement.

Reply:
we can check that later, it's a preliminary integration test :-)

Reply:
This has been removed for executing simple commands on the pool and is only used when acquiring a acquiring a connection out of the pool. You can see this as a protection layer between the physical connection and the API.